egg/core/language_analysis.py

```diff
@@ -124,3 +149,70 @@ def compute_distance(_list, distance):
 
         output_message = json.dumps(dict(topsim=topsim, epoch=epoch))
         print(output_message, flush=True)
+
+
+class PosDisent(Callback):
+    """
+    Positional disentanglement metric, introduced in "Compositionality and Generalization in Emergent Languages", 
+    Chaabouni et al., ACL 2020.
+    """
+    def __init__(self, print_train: bool = True, is_gumbel: bool = False):
+        super().__init__()
+        self.print_train = print_train
+        self.is_gumbel = is_gumbel
+
+    @staticmethod
+    def posdis(attributes, messages):
+        """
+        Two-symbol messages representing two-attribute world. One symbol encodes on attribute:
+        in this case, the metric should be maximized:
+        >>> samples = 1_000
+        >>> _ = torch.manual_seed(0)
+        >>> attribute1 = torch.randint(low=0, high=10, size=(samples, 1))
+        >>> attribute2 = torch.randint(low=0, high=10, size=(samples, 1))
+        >>> attributes = torch.cat([attribute1, attribute2], dim=1)
+        >>> messages = attributes
+        >>> PosDisent.posdis(attributes, messages)
+        0.9786556959152222
+        >>> messages = torch.cat([messages, torch.zeros_like(messages)], dim=1)
+        >>> PosDisent.posdis(attributes, messages)
+        0.9786556959152222
+        """
+        gaps = torch.zeros(messages.size(1))
+        non_constant_positions = 0.0
+
+        for j in range(messages.size(1)):
+            symbol_mi = []
+            h_j = None
+            for i in range(attributes.size(1)):
+                x, y = attributes[:, i], messages[:, j]
+                info = mutual_info(x, y)
+                symbol_mi.append(info)
+
+                if h_j is None:
+                    h_j = calc_entropy(y)
+
+            symbol_mi.sort(reverse=True)
+
+            if h_j > 0.0:
+                gaps[j] = (symbol_mi[0] - symbol_mi[1]) / h_j
+                non_constant_positions += 1
+
+        score = gaps.sum() / non_constant_positions
+        return score.item()
+
+    def print_message(self, logs: Interaction, tag: str, epoch: int):
+        message = logs.message.argmax(
+            dim=-1) if self.is_gumbel else logs.message
+
+        posdis = self.posdis(logs.sender_input, message)
+
+        output = json.dumps(dict(posdis=posdis, mode=tag, epoch=epoch))
+        print(output, flush=True)
+
+    def on_epoch_end(self, _loss, logs: Interaction, epoch: int):
+        if self.print_train:
+            self.print_message(logs, 'train', epoch)
+
+    def on_test_end(self, loss, logs, epoch):
+        self.print_message(logs, 'test', epoch)
```

Comment:
In the [topsim](https://github.com/facebookresearch/EGG/blob/master/egg/core/language_analysis.py#L103-L111) callback I made both (print_train and print_valid) optional but at least one must be specified. 



I'd say to implement both callbacks with the same feature for consistency. Both are fine with me

Reply:
Done